Sources/TensorFlow/Operators/NN.swift

```diff
@@ -293,279 +406,148 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
-        // closed form.
-        let value = maxPooled3D(kernelSize: kernelSize, strides: strides, padding: padding)
-        return (value, { v in
-            return Raw.maxPool3DGrad(
-                origInput: self,
-                origOutput: value,
-                grad: v,
-                ksize: [Int32(kernelSize.0), Int32(kernelSize.1), Int32(kernelSize.2),
-                        Int32(kernelSize.3), Int32(kernelSize.4)],
-                strides: [Int32(strides.0), Int32(strides.1), Int32(strides.2), Int32(strides.3),
-                          Int32(strides.4)],
-                padding: padding.raw
-            )
-        })
-    }
 
-    @inlinable
-    internal func _vjpAveragePooled2D(
-        kernelSize: (Int, Int, Int, Int),
-        strides: (Int, Int, Int, Int),
-        padding: Padding
-    ) -> (Tensor, (Tensor) -> Tensor) {
-        // TODO: Currently this is not higher order differentiable. Redefine in
-        // closed form.
-        let value = averagePooled2D(kernelSize: kernelSize, strides: strides, padding: padding)
-        return (value, { v in
-            Raw.avgPoolGrad(
-                origInputShape: self.shapeTensor,
-                grad: v,
-                ksize: [Int32(kernelSize.0), Int32(kernelSize.1),
-                        Int32(kernelSize.2), Int32(kernelSize.3)],
-                strides: [Int32(strides.0), Int32(strides.1),
-                          Int32(strides.2), Int32(strides.3)],
-                padding: padding.raw
-            )
-        })
-    }
+/// Computes a 2-D convolution using `self` as input, with the specified
+/// filter, strides, and padding.
+///
+/// - Parameters:
+///     - filter: The convolution filter.
+///     - strides: The strides of the sliding filter for each dimension of the
+///         input.
+///     - padding: The padding for the operation.
+/// - Precondition: `self` must have rank 4.
+/// - Precondition: `filter` must have rank 4.
 
-    @inlinable
-    internal func _vjpAveragePooled3D(
-        kernelSize: (Int, Int, Int, Int, Int),
-        strides: (Int, Int, Int, Int, Int),
-        padding: Padding
-    ) -> (Tensor, (Tensor) -> Tensor) {
-        // TODO: Currently this is not higher order differentiable. Redefine in
-        // closed form.
-        let value = averagePooled3D(kernelSize: kernelSize, strides: strides, padding: padding)
-        return (value, { v in
-            return Raw.avgPool3DGrad(
-                origInputShape: self.shapeTensor,
-                grad: v,
-                ksize: [Int32(kernelSize.0), Int32(kernelSize.1), Int32(kernelSize.2),
-                        Int32(kernelSize.3), Int32(kernelSize.4)],
-                strides: [Int32(strides.0), Int32(strides.1), Int32(strides.2), Int32(strides.3),
-                          Int32(strides.4)],
-                padding: padding.raw
-            )
-        })
-    }
+@differentiable(wrt: (x, filter), vjp: _vjpConv2D)
+func conv2D<Scalar: TensorFlowFloatingPoint>(
```

Comment:
Please make convXD, maxPoolXD, and avgPoolXD functions public.
```suggestion
public func conv2D<Scalar: TensorFlowFloatingPoint>(
```